### PR TITLE
update-vo-config v1.0

### LINKED ISSUE #20: VO Card API modified
As explained on the following broadcast: https://operations-portal.egi.eu/broadcast/archive/2838, the VO Card API has been modified. The [update-vo-config tool](https://github.com/quattor/tools/tree/master/update-vo-config) should be updated accordingly.

### PR DESCRIPTION
**Pull request by Pansanel**
The main changes of this new release are:
* Support of the new API of the [Operations Portail](https://operations-portal.egi.eu/)
* Removal of the old CIC_ID tag (not supported any more by the Operations Portal)
* Migration to Python 3
* Some pep8 fixes
* Fix #20

### CODE REVIEW

File: update-vo-config/setup.py [hunk 1]
```diff
@@ -1,4 +1,4 @@
-#!/usr/bin/env python
+#!/usr/bin/env python3
```
Comment (jouvin) [inline update-vo-config/setup.py]: @Pansanel thanks for maintaining this tool. Personnally I don't consider as a good practice to put `python3` rather tha `python` as `python3` is not always available,. In particular on Windows where there is no symlinks. For the the good practice is to use an `assert`. Something like:

```python
try:
    assert sys.version_info >= (3, 8)
except AssertionError:
    print("Error: Python 3.8+ required.")
    sys.exit(1)
```
(adjust the requirement to whatever you consider the appropriate value)

File: update-vo-config/update-vo-config [hunk 2]
```diff
@@ -1,4 +1,4 @@
-#! /usr/bin/env python
+#!/usr/bin/env python3
```
Comment (jouvin) [inline update-vo-config/update-vo-config]: Idem for all Python scripts

File: update-vo-config/setup.py (hunk 1 above)
Comment (Pansanel) [reply update-vo-config/setup.py]: Thanks, @jouvin for reviewing this PR. I have updated the scripts as suggested. The tool need to load urllib.request, that is only available on Python 3. In consequence, I have linked the ImportError with the use of Python 2.

**Comment by jouvin** [comment]
Thanks. LGTM. I will try to test it before merging it...

**Comment by jouvin** [comment]
I started to test it but I'm waiting for my token... I'd add to the doc the Conda-based installation which for me is attractive if you use SCDB: this way you just copy the script in your SCDB (`utils/` directory), without executing the `setypu.py` script, and don't depend on the system-installation of Python: you just configure your Conda environement with the required module. It should be no more than the following commands, once you have installed Anaconda or MiniConda :

```bash
conda create -n quattor urllib3 openssl
conda activate quattor
```

I'll update you as soon as I got my token and have been able to run the script.

**Comment by jouvin** [comment]
@pansanel I ran the script but it crashed when updating VOM server `voms.ipb.ac.rs`. The traceback is:

```
No certicate defined for voms.ipb.ac.rs voms server.
Updating template for VOMS server voms.ipb.ac.rs (C:/Users/jouvin/AppData/Local/Temp\vo/certs\voms.ipb.ac.rs.pan)
Traceback (most recent call last):
  File "C:\Users\jouvin\PycharmProjects\Quattor\tools\update-vo-config\update-vo-config", line 603, in <module>
    update_vo_config()
  File "C:\Users\jouvin\PycharmProjects\Quattor\tools\update-vo-config\update-vo-config", line 599, in update_vo_config
    update_voms_servers(vo_card_data, basedir, namespaces)
  File "C:\Users\jouvin\PycharmProjects\Quattor\tools\update-vo-config\update-vo-config", line 405, in update_voms_servers
    if old_x509_cert:
UnboundLocalError: local variable 'old_x509_cert' referenced before assignment
```

Can you reproduce it? I ran the script on a brand new base dir.

**Comment by Pansanel** [comment]
@jouvin : I have never seen this issue before, but looking at the code, I cannot figure out why, as the variable could effectively be referenced before its assignment. I have fixed this issue in the commit c77c4fd.

**Comment by jouvin** [comment]
@Pansanel thanks, it works fine now! Tested on Windows. Could you add the conda-based installation instructions to the README?

@ned21 @jrha Jérome added his script to this repo and in principle I find it a good idea, as this script is useful for anybody running an EGI grid site, no matter whether SCDB or Aquilon is used. But the structure of the repo is still too much Aquilon-oriented and I think it is not optimal. The main `INSTALL` file applies only to Aquilon tools that are in the `src` directory when the new tool is in `update-vo-config` directory.  What about restructuring everything with top-level directories like `aquilon`, `grid` and moving everything except `update-vo-config` in `aquilon/` and `update-vo-config` in `grid`?

**Comment by Pansanel** [comment]
@jouvin I have added the conda installation instructions.

**Comment by jose-caballero** [comment]
Hi,

trying the script, it failed for VOs cic.testworkflow.fr and vo.stars4all.eu, with a traceback like this:

```
Traceback (most recent call last):
  File "./update-vo-config", line 589, in <module>
    update_vo_config()
  File "./update-vo-config", line 564, in update_vo_config
    vo_card_data = get_vo_card_data(VO_CARD_URI)
  File "./update-vo-config", line 548, in get_vo_card_data
    vo_card_data.append(parse_vo_card(vocard))
  File "./update-vo-config", line 513, in parse_vo_card
    voms_endpoint = parse_voms_server_data(voconfig['name'], server)
  File "./update-vo-config", line 474, in parse_voms_server_data
    hostname = hostname.text.strip()
AttributeError: 'NoneType' object has no attribute 'strip'
```



Wrapping the line https://github.com/quattor/tools/blob/master/update-vo-config/update-vo-config#L537 in a try - except block prevents it from aborting. 


Cheers,
Jose

File: update-vo-config/update-vo-config [hunk 3]
```diff
@@ -341,23 +354,25 @@ def extract_old_cert(server_cert_tpl=None, debug=False):
...
+            )
         else:
             for server in vo_config['voms_endpoints']:
                 if server.hostname not in voms_servers:
                     voms_servers[server.hostname] = server.cert
     for (hostname, cert) in voms_servers.items():
         server_cert_ns = os.path.join(namespaces['cert_dir'], hostname)
         server_cert_tpl = os.path.join(basedir, server_cert_ns + TPL_SUFFIX)
+        x509_cert = None
+        old_X509_cert = None
```
Comment (jrha) [inline update-vo-config/update-vo-config]:
```suggestion
        old_x509_cert = None
```

**Comment by jouvin** [comment]
I merge this PR that was on hold partly because of my remark about the directory structure which is an issue worth another issue/PR.

**Pull request merged**